xscen/aggregate.py

```diff
@@ -122,43 +214,170 @@ def climatological_mean(
+            if op == "mean" and ds_has_std:
+                for vv in ds_rolling.data_vars:
+                    if (
+                        "std" in vv
+                        or "standard deviation"
+                        in ds_rolling[vv].attrs["description"].lower()
+                    ):
+                        ds_rolling[vv] = np.sqrt(ds_rolling[vv])
+
+            # Select the windows at provided stride, starting from the first full window's operation result
+            ds_rolling = ds_rolling.isel(year=slice(window - 1, None, stride))
+
+        elif op == "linregress":
+
+            def _ulinregress(x, y, **kwargs):
+                # Wrapper for scipy.stats.linregress to unpack multiple return values in xr.apply_ufunc
+                valid_x = ~np.isnan(x)
+                valid_y = ~np.isnan(y)
+                mask = valid_x & valid_y
+                if np.sum(mask) >= kwargs.get("min_periods", 1):
+                    reg = scipy.stats.linregress(
+                        x, y, alternative=kwargs.get("alternative", "two-sided")
+                    )
+                    out = np.array(
+                        [
+                            reg.slope,
+                            reg.intercept,
+                            reg.rvalue,
+                            reg.pvalue,
+                            reg.stderr,
+                            reg.intercept_stderr,
+                        ]
+                    )
+                else:
+                    out = np.full(6, np.nan)
+                return out
+
+            # prepare kwargs
+            linreg_kwargs = {
+                k: v for k, v in op_kwargs.items() if "keep_attrs" not in k
+            }
+            linreg_kwargs["min_periods"] = min_periods
+
+            # unwrap DatasetRolling object and select years subset
+            dsr_construct = ds_rolling.construct(window_dim="window", keep_attrs=True)
+            dsr_construct = dsr_construct.isel(year=slice(window - 1, None, stride))
+
+            # construct array to use years as x values (==regressors) in xr.apply_ufunc
+            years_as_x_values = xr.DataArray(
+                np.arange(dsr_construct.window.size)
+                .repeat(dsr_construct.year.size)
+                .reshape(dsr_construct.window.size, dsr_construct.year.size)
+                + dsr_construct.year.values
+                - window
+                + 1,
+                dims=["window", "year"],
+                coords={
+                    "window": dsr_construct.window.values,
+                    "year": dsr_construct.year.values,
+                },
+            )
+
+            # apply linregress along windows
+            ds_rolling = xr.apply_ufunc(
+                _ulinregress,
+                years_as_x_values,
+                dsr_construct,
+                input_core_dims=[["window"], ["window"]],
+                output_core_dims=[["linreg_param"]],
+                vectorize=True,
```

Comment:
`vectorize` is almost always slower. There might be a way to rewrite `_ulinregress` so that we don't need `vectorize`... If you find this to be very slow for your use case, it might be worth digging into this.

Reply:
Noted.
I would have to dig into doing this without the vectorization.
It is already quite fast. Parallelization already makes it much faster than by iteration over the grid.